polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,134 @@
+
+        for polling_station_id, uprns in polling_station_lookup.items():
+            uprns_in_council.filter(uprn__in=uprns).update(
+                polling_station_id=polling_station_id
+            )
+
+
+class StationSet(CustomSet, AssignPollingStationsMixin):
+    NotImplemented
+
+
+class DistrictSet(CustomSet, AssignPollingStationsMixin):
+    NotImplemented
+
+
+class AddressList(AssignPollingStationsMixin):
+    def __init__(self, logger):
+        self.elements = []
+        self.logger = logger
+
+    def append(self, address):
+
+        if (
+            not address["address"]
+            or not address["postcode"]
+            or not address["uprn"]
+            or not address["council"]
+            or not address["polling_station_id"]
+        ):
+            self.logger.log_message(
+                logging.DEBUG,
+                "Record with empty required fields found:\n%s",
+                variable=address,
+                pretty=True,
+            )
+            return
+
+        self.elements.append(address)
+
+    def get_uprn_lookup(self):
+        # for each address, build a lookup of uprn -> set of station ids
+        uprn_lookup = {}
+        for record in self.elements:
+            uprn = record["uprn"]
+            if not uprn:
+                continue
+            if uprn in uprn_lookup:
+                uprn_lookup[uprn].add(record["polling_station_id"])
+            else:
+                uprn_lookup[uprn] = {record["polling_station_id"]}
+
+        return uprn_lookup
+
+    # TODO be more clever to report on duplicates.
+    def remove_duplicate_uprns(self):
+        uprn_lookup = self.get_uprn_lookup()
+        self.elements = [
+            record for record in self.elements if len(uprn_lookup[record["uprn"]]) == 1
+        ]
+
+    def get_polling_station_lookup(self):
+        # for each address, build a lookup of polling_station_id -> set of uprns
+        polling_station_lookup = {}
+        for record in self.elements:
+            if record["polling_station_id"] in polling_station_lookup:
+                polling_station_lookup[record["polling_station_id"]].add(record["uprn"])
+            else:
+                polling_station_lookup[record["polling_station_id"]] = set(
+                    record["uprn"]
+                )
+
+        return polling_station_lookup
+
+    def remove_records_not_in_addressbase(self, addressbase_data):
+        self.elements = [e for e in self.elements if e["uprn"] in addressbase_data]
+
+    def remove_records_that_dont_match_addressbase(self, addressbase_data):
+        for input_record in self.elements:
+            addressbase_record = addressbase_data[input_record["uprn"].lstrip("0")]
+
+            if (
+                Postcode(input_record["postcode"]).with_space
+                == Postcode(addressbase_record["postcode"]).with_space
+            ):
+                continue
+            else:
+                self.elements.remove(input_record)
```

Comment:
Have you got a handle on the treadeoff here? i.e: what sort of error rate this gives us on a really messy dataset?

My instinct is that although this will simplify things a lot on the code side and makes the approach simple to explain, this is probably still going to require quite a bit of the manual overhead on each import script to spot the false-positives.

Does that reflect your experience?

Reply:
I think your hunch is probably right. My ideal would be able to switch on a 'strict' mode which enforces address matching. In most LAs I think this would be the default and cut down on the errors. However there are some (Fylde springs to mind) where dropping the address matching stage allows us to get much better coverage (think 40%->70% in case of Fylde) in which case checking for the false positives would probably be worth it.
I had left it 'unstrict' for now as I felt like it was more useful to see if the errors seemed any different to before, but tbh I've not noticed much.

Reply:
I think one of the big things I'd hope to achieve here is reducing the amount of manual checks required on each import script. If we could massively reduce the overhead on each area and the cost is that there are some councils with really messy data where we only get 40% coverage that seems like a better tradeoff.. but it is your call which way to go with it at this point. Do you think you will look at adding 'strict mode' this side of the 2021 campaign cycle, or do you think you will try an election without it and see what happens?

Reply:
Agree about trying to have fewer manual checks required. There'll be some sort of address matching (either as a flag or just for everything) before 2021 imports.